modules/dhcp/subnet_service.rb

```diff
@@ -0,0 +1,105 @@
+class Proxy::DHCP::SubnetService
+  def self.instance_with_default_parameters
+    Proxy::DHCP::SubnetService.new(Proxy::MemoryStore.new, Proxy::MemoryStore.new,
+                                   Proxy::MemoryStore.new, Proxy::MemoryStore.new,
+                                   Proxy::MemoryStore.new, Proxy::MemoryStore.new)
+  end
+
+  # rubocop:disable Metrics/ParameterLists
+  def initialize(subnets_store, leases_ip_store, leases_mac_store, reservations_ip_store, reservations_mac_store,
+      reservations_name_store)
+    @subnets = subnets_store
+    @leases_by_ip = leases_ip_store
+    @leases_by_mac = leases_mac_store
+    @reservations_by_ip = reservations_ip_store
+    @reservations_by_mac = reservations_mac_store
+    @reservations_by_name = reservations_name_store
+  end
+  # rubocop:enable Metrics/ParameterLists
+
+  def add_subnet(subnet)
+    raise Proxy::DHCP::Error, "Unable to Add Subnet #{subnet}" if find_subnet(subnet.network)
+    @subnets[subnet.network] = subnet
+  end
+
+  def add_subnets(*subnets)
+    subnets.each { |s| add_subnet(s) }
+    subnets
+  end
+
+  def delete_subnet(subnet_address)
+    @subnets.delete(subnet_address)
+  end
+
+  def find_subnet(address)
+    to_ret = @subnets[address]
+    return to_ret if to_ret # we were given a subnet address
+
+    # TODO: this can be done much faster
+    @subnets.values.each do |subnet|
+      return subnet if subnet.include?(address)
+    end
+
+    nil
+  end
+
+  def all_subnets
+    @subnets.values
+  end
+
+  def add_lease(subnet_address, record)
+    @leases_by_ip[subnet_address, record.ip] = record
+    @leases_by_mac[subnet_address, record.mac] = record
+  end
+
+  def add_host(subnet_address, record)
```

Comment:
Could we have some debug or trace level logging in these methods?  I think it might be useful - we used to in add_record.